### ISSUE TITLE
Race with numericInput (and renderUI)

### ISSUE THREAD

**Issue by smartinsightsfromdata** [rstudio/shiny#787]
My app uses numericInput, within a renderUI context.
(Sadly) I can easily generate a race, where shiny gets into a never ending loop.
It is worth noting that the numericInput are programmatically sent to the client using renderUI.

I have already reported a similar problem (i.e. a race), happening using DT & the new checkboxes  https://github.com/rstudio/DT/issues/35

As it seems difficult to reproduce the problem (or it does not happen with every machine?) 
I've posted a video in a shared folder (in the same folder there is also the DT race).

I wonder if there is a general method to avoid, prevent or resolve this type of issues?

The two videos are here: https://www.dropbox.com/l/tpzeafZeyB2JOas3PrusKs/invite 
or
https://www.dropbox.com/sh/eevmfrahay9bq3s/AADu5qSWu81OAlVISSTndTCva?dl=0

**Comment by jcheng5** [comment]
Thanks for the report! We'd need to see the code to even begin to help you with this. The other thing you can try is to use the reactlog, see `?showReactLog` and carefully follow the instructions there. In particular, restart the R process before each time you are going to use it on an app.

**Comment by smartinsightsfromdata** [comment]
@jcheng5 This is a code that can easily generate a race between client and server (please let me know if would be helpful  to do another screen movie for this specific example).  
So far I've seen on my laptop (2 core Macbook pro 3 GHz Intel Core i7) and on windows and apple clients against shiny server. 
(Of course this is bound to happen more with shiny server and a remote client...)

Please note that there is *no way* to resolve the race beside *killing the app*.

It would be great if you could kindly:
- indicate how to avoid these issues
- potentially provide a method to "debounce" the client (or to enforce a closer connection in certain cases like a sort of ```commit```?).
Incidentally, I tried to use the reactlog, but it is very difficult to read, as so many states are generated when the loop set in.
I will paste below also the example with datatable.

```R
rm(list=ls()) 
#####
library(DT)
library(shiny)
# options(shiny.reactlog=TRUE)

id <- 1:10
label <- paste0("lab","-",id)
ratio <- runif(10, 1.0, 10.0)
ratio <- round(ratio,0)
df <- ds <- data.frame(id=id, label=label,ratio=ratio, stringsAsFactors = FALSE)
ds <- rbind(ds, c(11,"RestTo100",0))

ui <- fluidPage(
	fluidRow(
		column(5, uiOutput('resSli')) ,
column(7, textOutput("resAv",inline=TRUE),
          plotOutput('pie'))
		)
	)

server <- function(session, input, output) {
cache <- reactiveValues()
cache$ds <- ds

output$resSli <- renderUI({

if(is.null(resRDs()) )  return()

isolate({
df_  <- resRDs()
buttn <- 1

iter <- length(df_$id) 

  iter <- iter / 2
  toRender <- lapply(1:iter, function(i) {
            fluidRow(
            column(2,h6(df_$label[i], 
      style = "font-family: 'Helvetica Neue', small;
        font-weight: 200; line-height: 1.0;")),
            column(3,numericInput(paste0(df_$label[i],buttn),
             label = NULL, value = df_$ratio[i],min = 1, max = 100)),
            column(2,h6(df_$label[i+iter], 
      style = "font-family: 'Helvetica Neue', small;
        font-weight: 200; line-height: 1.0;")),
            column(3,numericInput(paste0(df_$label[i+iter],buttn),
             label = NULL, value = df_$ratio[i+iter], min = 1, max = 100))
            )
            })
  return(toRender)

})
})
######

resRDs <- reactive({


isolate({
df_ <-ds_ <- cache$ds

loop <- length(df_$id) -1
buttn <- 1
})

valLabs <- sapply(1:loop, function(i) {
if(!is.null( input[[paste0(df_$label[i],buttn)]] )) {
      as.numeric(input[[paste0(df_$label[i],buttn)]])  
  }
  })

if( loop == sum(sapply(valLabs,length)) && loop>0){

for (i in 1:loop) {

if( is.na(valLabs[i])) valLabs[i] <- as.numeric(ds_$ratio[i])
if(valLabs[i]<0 ) valLabs[i] <- as.numeric(ds_$ratio[i])
}

df_$ratio[1:loop] <- valLabs
ds_$ratio[1:loop] <- valLabs

}

resAv <- 100 - sum(as.numeric(df_$ratio) )
if( resAv < 0) resAv <- 0

ds_[ds_$label=="RestTo100",   "ratio"]  <- resAv
cache$ds <- ds_

return(df_)
})

#########

output$resAv <- renderText({ 
if( is.null( cache$ds )) return()
ds_ <- cache$ds
totsAv <- round(as.numeric(ds_[ds_$label=="RestTo100", "ratio"]),0)
return(paste0("Available resources: ",totsAv,"%"))
})

##########



output$pie <- renderPlot({
if( is.null(resRDs())) return()
df_ <- cache$ds
df_$ratio <- round(as.numeric(df_$ratio),0)

lbls <- paste(df_$label, df_$ratio) # add percents to labels 
lbls <- paste(df_$label,"%",sep="") # ad % to labels 
pie(df_$ratio,labels = lbls, col=rainbow(length(lbls)),
  	main="Pie Chart of Labels")
})


}


shinyApp(ui, server)
```
This is the datatable example (already posted in the datatable github repository:
```R

rm(list=ls()) 
#####
library(DT)
library(shiny)
options(shiny.reactlog=TRUE)
quantity <- id <- 1:nrow(iris)
label <- paste0("lab","-",quantity)
iris_ <- cbind(id=id,quantity=quantity,label=label,iris ,stringsAsFactors = FALSE)
cols <- colnames(iris_)

ui <- fluidPage(
	fluidRow(
		column(8,DT::dataTableOutput('demTb')),
		column(3,uiOutput("demSli"))
		)
	)

server <- function(session, input, output) {


output$demSli <- renderUI({

if(is.null(input$demTb_selected) ) return()
  
isolate({
index <- input$demTb_selected

labs <- iris_$label[index] 
pages <- "test"
iter <- length(labs)
buttn <- 1
valLabs <- sapply(1:iter, function(i) {
if(is.null(input[[paste0(pages,"d",labs[i],buttn)]] )) {
          0
} else {  as.numeric(input[[paste0(pages,"d",labs[i],buttn)]])  }
}) 
#
toRender <- lapply(1:iter, function(i) {
  sliderInput(inputId = paste0(pages,"d",labs[i],buttn),
              label =  h6(paste0(labs[i],"")),
              min = -100,
              max = 100,
              step = 1,
              value = valLabs[i],
              post="%",
              ticks = FALSE, animate = FALSE)
              })
})
      return(toRender)

})
#
rds <- reactive({
if( is.null(input$demTb_selected) ) return(iris_)
index <- input$demTb_selected
isolate({
labs <- iris_$label[index] 
pages <- "test"
iter <- length(labs)
buttn <- 1
}) # end isolate
valLabs <- sapply(1:iter, function(i) {
    if(is.null(input[[paste0(pages,"d",labs[i],buttn)]] )) {
      0
    } else {  
      as.numeric(input[[paste0(pages,"d",labs[i],buttn)]])/100  
    }
  })
  
  df_ <- data.frame(label=labs, multi=valLabs, stringsAsFactors = FALSE)
  df_ <- merge(iris_,df_,by="label", all.x=T)
  ifelse( is.na( df_$multi), df_$quantity, df_$quantity*(1 + df_$multi) )
  df_$multi <- NULL
  return(df_)
  }) 


output$demTb = DT::renderDataTable({


if(is.null(rds() )) return()
df_ <- rds() 
df_ <- df_[cols]
df_ <- df_[with(df_,order(as.numeric(id))),]

DT::datatable(df_,extensions= 'Scroller', rownames = checkboxRows(df_, input$demTb_selected),
      escape = -1,
      options=list(
      # deferRender = TRUE,
      dom = 'frtiS',
      scrollX = TRUE,
      scrollY = 300,
      scrollCollapse = TRUE,
      searching=F
      ))

    })

}


shinyApp(ui, server)
```

**Comment by smartinsightsfromdata** [comment]
As posted here https://github.com/rstudio/DT/issues/35 so far I have tried minor changes with no success.

Could somebody kindly spread some light?

**Comment by jcheng5** [comment]
For the first example you posted (with the pie chart), I would strongly recommend moving away from the circular dependency that `resRDs` and `resSli` have on each other. As far as I can tell, resRDs is:

1. Replacing any `NA` values with the previous good value, and
2. Filling in the "RestTo100" value dynamically.

I would do both of these using an observer and `updateNumericInput` (and then, only if the value has actually changed).

**Comment by jcheng5** [comment]
I can't get the second example to render anything, I just get a blank screen...? I just installed `jsonlite` from CRAN and the latest versions of `htmlwidgets` and `DT` from github.

**Comment by yihui** [comment]
@jcheng5 I think you need the Github version of shiny as well. That might be the reason. If it is not, see if you have got any errors in the JavaScript console.

**Comment by smartinsightsfromdata** [comment]
@jcheng5 thanks for the help so far.  Few more questions to properly tackle this problem:

At very high level, this should be a typical pattern with data entry, which is the common thread between the two examples.
Basically I try to:
1. source data, typically with a reactive (e.g. resRDs) from a db (I simplified this part in the code above) 
2. use a data entry widget that also displays data  (numericInput or the checkbox feature in DT)
3. perform an action that does have an effect on the entry widget. In the first example the action is to render a slider on the screen.

I understand your point about using ```updateNumericInput``` in the pie chart example. Here the problem is that I don't know how many data entry I need for each session: they can vary.  I suppose I can use ```renderUI``` in conjunction with ```updateNumericInput```  and hopefully this will eliminate the loop in this example: I will try.

The first case is possibly more complex.  I have adopted  to render sliders with ```renderUI``` and ```sliderInput``` vs.  ```updateSliderInput```  as I don't know which row of the table the user will want to change (the checkbox will tell me), nor if many rows need to be changed at the same time.

Clearly to have these "tasks" in separate functions creates the potential for loops (I should get a prize for the most loops created!), but I am not sure how to go about combining these functions to avoid them.

Should I just wrap the functions with observe (if I understand your suggestion)?
* observe
  * reactive (to get data)
  * render data (using DT or numericInput)
  * renderUI (to render a sliderInput or and updateSliderInput)
* end of observe

**Comment by jcheng5** [comment]
OK, I took a much closer look at your pie example, and in this case the circular reference isn't needed at all. I assumed that editing input 1 could cause input 2 to be rewritten or something, but that seems not to be the case.

So the problem here is that `resRDs` (what does that stand for, anyway?) is going to change in response to user input. You've taken a dependency on it in the `renderUI`, even though you don't actually want the UI to be re-rendered when user input changes. Breaking the `resRDs` into multiple reactives helps solve the problem here.

Here's my solution:
https://gist.github.com/jcheng5/33851fda3dd6078ba575

You can click on "Revisions" on the right to see the changes I made. Besides turning `resRDs` into separate reactives (but still keeping the `resRDs` reactive--just putting it together from the other reactives), I also got rid of the `df`/`ds` distinction (oops, missed one reference in line 9, ignore that) and got rid of the `cache` reactiveValues. I'm not sure what purpose that cache served in the first place.

Note that `renderUI` now depends on `rawData()`, so it doesn't rewrite the numeric inputs when their values change--that circular reference is gone.

So I guess the main takeaway, is to make sure that when you use `renderUI` (especially to render inputs), make sure it only takes the reactive dependencies that are absolutely necessary. (This makes for a better user experience anyway, as your original version of the app caused the numeric input to lose focus whenever it was changed.)

**Comment by smartinsightsfromdata** [comment]
@jcheng5 Your code for numericInput resolve the issue and works like a charm :)

On the other hand I haven't made any progress on the endless loop with DT :(

Could you kindly have a look at that also?

**Comment by cmdcolin** [comment]
I had a very similar problem to that described in this thread, where the app would go into an infinite loop occaisionally when changing an input value (especially if changing it quickly). My app used a similar setup with components created via renderUI and too. 

Basically this sentence from @jcheng5 helped me solve it

"You've taken a dependency on it [a reactive value] in the renderUI, even though you don't actually want the UI to be re-rendered when user input changes"

So basically I had mistakenly used a reactive value for the initial values used by the component. It was pretty easy to remove this dependency in my case, but I appreciate this thread for providing that context :)
